info: Add --system option to show system info

kangtegong opened the pull request:
In order to get kernel information with uftrace without any problem, I think /proc/sys/kernel/perf_event_paranoid should be 2 or less. But many distros nowadays set this value to more than 3 as default.

I wanted to inform users about this in the docs, but it's so general that it's unclear which docs to leave this information in.

So this commit suggests checking this information
when configuring uftrace.

Closes: #1424

kangtegong committed:
info: Add --system option to show system info

Some kernel values affect what uftrace can do (most notably
perf_event_paranoid), but there was no easy way to check them.

So this commit adds Add --system to print basic system info:

Usage:
    $ uftrace info --system
    cpu info            : AMD EPYC 7313P 16-Core Processor
    number of cpus      : 32 / 32 (online / possible)
    memory info         : 70.0 / 125.7 GB (free / total)
    kernel version      : Linux 5.15.0-139-generic
    hostname            : myserver
    distro              : "Ubuntu 20.04.6 LTS"
    perf_event_paranoid : 4
    kptr_restrict       : 1

Closes: #1424

Signed-off-by: Kang Minchul <tegongkang@gmail.com>

kangtegong commented:
If there's any better idea, please let me know.

namhyung commented:
Showing the value when `configure` runs is not ideal because it may be far from users actually run uftrace.

I think it'd be better if `uftrace info` could show such information.  Maybe we can have `--system` option to display various system information related to uftrace like this one.  Additionally it can display the same thing in the "system information" section in the `uftrace info` output.  Something like this?

```
$ uftrace info --system
cpu info            : 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
number of cpus      : 8 / 8 (online / possible)
memory info         : 4.3 / 31.0 GB (free / total)
kernel version      : Linux 6.1.25-1rodete1-amd64
hostname            : bangji
distro              : "Debian GNU/Linux rodete"
perf_event_paranoid : 3
kptr_restrict       : 1
```

honggyukim commented:
I think `uftrace info --features` might be better and we can show all the feature info in this option as follows.

I would like it to have compile time and run time feature tests separately.

kangtegong commented:
Thank you for your detailed feedbacks, 
I will send you a revised PR reflecting your feedback shortly :)

namhyung commented:
Maybe `--features` can be for compile time feature tests then.  But I'm not sure how much difference it can get from `uftrace --version`.

kangtegong commented:
Hello :) I'm sorry for the long delay — I'm coming back to finish this work.
Before submitting the revised PR, I'd like to confirm the expected output format.

Does this look like the right direction? 
The `--system` option is intended to show runtime system information relevant to uftrace (such as perf_event_paranoid), and the same fields would also be included in the existing "system information" section of uftrace info.


```
$ uftrace info --system
cpu info            : AMD EPYC 7313P 16-Core Processor
number of cpus      : 32 / 32 (online / possible)
memory info         : 80.0 / 125.7 GB (free / total)
kernel version      : Linux 5.15.0-139-generic
hostname            : gpuserver
distro              : "Ubuntu 20.04.6 LTS"
perf_event_paranoid : 4
kptr_restrict       : 1
```

```
$ uftrace info
# system information
# ==================
# program version     : v0.11-812-g0270 ( x86_64 dwarf python3 luajit tui perf sched dynamic kernel )
# recorded on         : Wed Apr 15 10:52:01 2026
# cmdline             : uftrace record ./abc
# cpu info            : AMD EPYC 7313P 16-Core Processor
# number of cpus      : 32 / 32 (online / possible)
# memory info         : 80.0 / 125.7 GB (free / total)
# system load         : 2.12 / 2.14 / 2.17 (1 / 5 / 15 min)
# kernel version      : Linux 5.15.0-139-generic
# hostname            : gpuserver
# distro              : "Ubuntu 20.04.6 LTS"
# perf_event_paranoid : 4
# kptr_restrict       : 1
#
# process information
# ===================
# number of tasks     : 1
# task list           : 3888990(abc)
# exe image           : /home/minchul/uftrace/abc
# build id            : 7dcebdb0197027aa08a000582c780fa864c53e64
# pattern             : regex
# exit status         : exited with code: 0
# elapsed time        : 0.004318602 sec
# cpu time            : 0.004 / 0.000 sec (sys / user)
# context switch      : 2 / 0 (voluntary / involuntary)
# max rss             : 4924 KB
# page fault          : 1 / 363 (major / minor)
# disk iops           : 0 / 8 (read / write)
```

namhyung commented:
Looks good.

kangtegong commented:
Thank you. I’ve just sent the updated PR along with the revised commit message.
Please feel free to let me know if you have any comments :)
Have a great day!

namhyung commented on cmds/info.c:
```diff
@@ -894,6 +894,53 @@ static int read_utc_offset(void *arg)
+	fp = fopen("/proc/sys/kernel/perf_event_paranoid", "r");
+	if (fp != NULL) {
+		if (fscanf(fp, "%d", &paranoid) != 1)
+			paranoid = -1;
```
`-1` for `perf_event_paranoid` is a valid option, so you need to choose a different value.

namhyung commented on cmds/dump.c:
```diff
@@ -554,7 +554,7 @@ static void dump_raw_header(struct uftrace_dump_ops *ops, struct uftrace_data *h
 				   "CPUINFO",	   "MEMINFO",	   "OSINFO",	  "TASKINFO",
 				   "USAGEINFO",	   "LOADINFO",	   "ARG_SPEC",	  "RECORD_DATE",
-				   "PATTERN_TYPE", "VERSION",	   "UTC_OFFSET" };
+				   "PATTERN_TYPE", "VERSION",	   "UTC_OFFSET",  "KERNELINFO" };
```
I'd say `SYSCTL` instead of `KERNELINFO` as in this command.

```
$ sysctl kernel.perf_event_paranoid kernel.kptr_restrict
kernel.perf_event_paranoid = 2
kernel.kptr_restrict = 0
```

But I'm ok with reading the value from sysfs.

namhyung commented on cmds/info.c:
```diff
@@ -894,6 +894,53 @@ static int read_utc_offset(void *arg)
+	fp = fopen("/proc/sys/kernel/kptr_restrict", "r");
+	if (fp != NULL) {
+		if (fscanf(fp, "%d", &kptr) != 1)
+			kptr = -1;
```
`kptr_restrict` looks fine though.

namhyung commented on cmds/info.c:
```diff
@@ -894,6 +894,53 @@ static int read_utc_offset(void *arg)
+	struct uftrace_info *info = &handle->info;
+	char *buf = rha->buf;
+
+	info->perf_event_paranoid = -1;
```
Here as well.

namhyung commented on cmds/info.c:
```diff
@@ -1067,6 +1116,14 @@ void process_uftrace_info(struct uftrace_data *handle, struct uftrace_opts *opts
 	}
 
+	if (info_mask & KERNELINFO) {
+		if (info->perf_event_paranoid >= 0)
```
Should be fixed.

namhyung commented on cmds/info.c:
```diff
@@ -1255,11 +1312,57 @@ static void print_task_info(struct uftrace_data *handle)
+	pr_out(fmt, "kernel version", info->kernel);
+	pr_out(fmt, "hostname", info->hostname);
+	pr_out(fmt, "distro", info->distro);
+	if (info->perf_event_paranoid >= 0)
```
And here too.

namhyung reviewed (commented):
In general, looks fine.  Just a few new nits.

kangtegong commented:
Thanks for your comment :) 
I amended the PR. 

I renamed `KERNELINFO` to `SYSCTL` across dump.c, info.c, and uftrace.h,
and switched the `perf_event_paranoid` sentinel from -1 to `INT_MIN.`

namhyung reviewed (approved):
LGTM
